File: java/data-structure/Array/src/main/java/ru/reimu/array/Array.java
package ru.reimu.array;

/**
 * @Author: Tomonori
 * @Date: 2020/2/13 14:49
 * @Title:
 * @Desc: ↓ ↓ ↓ ↓ ↓
 * -----
 */
public class Array<T> {

    private T[] data;
    private int   size;

    public Array() {
        this(10);
    }

    /**
     * 传入数组的容量capacity构造Array
     * @param capacity
     */
    public Array(int capacity) {
        this.data = (T[]) new Object[capacity];
        this.size = 0;
    }

    /**
     *
     * @return 元素个数
     */
    public int getSize() {
        return size;
    }

    /**
     *
     * @return 数组的容量
     */
    public int getCap() {
        return data.length;
    }

    public T get(int index) {
        if (index < 0 || index >= size) {
            throw new IllegalArgumentException("Get failed, Require index >= 0 and index <= size");
        }
        return data[index];
    }

    public T getFirst() {
        return get(0);
    }

    public T getLast() {
        return get(size - 1);
    }

    public void set(int index, T e) {
        if (index < 0 || index >= size) {
            throw new IllegalArgumentException("Set failed, Require index >= 0 and index <= size");
        }
        data[index] = e;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * 查找数组中是否有元素e
     * @param e
     * @return
     */
    public boolean contains(T e) {
        for (T ele : data) {
            //值比较
            if (ele.equals(e)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 查找数组中的元素索引，-1没有找到
     * @param e
     * @return
     */
    public int find(T e) {
        for (int i = 0; i < size; i++) {
            if (data[i].equals(e)) {
                return i;
            }
        }
        return -1;
    }

    /**
     * 所有元素钱添加元素
     * @param e
     */
    public void addFirst(T e) {
        add(0, e);
    }

    /**
     * 所有元素后添加元素
     * @param e
     */
    public void addLast(T e) {
        add(size, e);
    }

    /**
     * 在第index的位置插入一个新元素
     * @param index
     * @param e
     */
    public void add(int index, T e) {
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("Add failed, Require index >= 0 and index <= size");
        }

        if (size == data.length) {
            resize(2 * data.length);
        }

        for (int i = size - 1; i >= index; i--) {
            data[i + 1] = data[i];
        }

        data[index] = e;
        size++;
    }

    private void resize(int newCapacity) {
        T[] newArr = (T[]) new Object[newCapacity];

        for (int i = 0; i < size; i++) {
            newArr[i] = data[i];
        }
        data = newArr;
    }

    /**
     * 从数组删除index位置的元素，返回删除的元素
     * @param index
     * @return
     */
    public T remove(int index) {
        if (index < 0 || index >= size) {
            throw new IllegalArgumentException("Remove failed, index is illegal");
        }

        T ele = data[index];
        for (int i = index + 1; i < size; i++) {
            data[i - 1] = data[i];
        }
        size--;
        data[size] = null;

        //防止减容resize过于着急(Eager)，在1/4的时候才减容
        //!=0防止数组中只有一个的情况
        if (size == data.length / 4 && data.length / 2 != 0) {
            resize(data.length / 2);
        }

        return ele;
    }

    public T removeFirst() {
        return remove(0);
    }

    public T removeLast() {
        return remove(size - 1);
    }

    /**
     * 删除数组中的某个元素e
     * @param e
     */
    public void removeElement(T e) {
        int index = find(e);

        if (index != -1) {
            remove(index);
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("Array: size = %d, capacity = %d\n", size, data.length));
        sb.append('[');

        for (int i = 0; i < size; i++) {
            sb.append(data[i]);
            if (i != size - 1) {
                sb.append(", ");
            }
        }

        sb.append(']');

        return sb.toString();
    }
}

File: golang/data-structure/array/array/array.go
package array

import (
	"bytes"
	"errors"
	"fmt"
)

/**
 * @Author: Tomonori
 * @Date: 2020/2/14 12:27
 * @Title:
 * --- --- ---
 * @Desc:
 */
func NewArray() (object *Array) {
	return NewArrayWithCap(10)
}

func NewArrayWithCap(cap int) (object *Array) {
	object = &Array{
		data: make([]interface{}, cap),
		cap:  cap,
	}
	return
}

type Array struct {
	data []interface{}
	size int
	cap  int
}

func (a *Array) GetSize() int {
	return a.size
}

func (a *Array) GetCap() int {
	return len(a.data)
}

func (a *Array) Get(index int) (ele interface{}, err error) {
	err = a.requireIndexFailed(index)
	ele = a.data[index]
	return
}

func (a *Array) Set(index int, val interface{}) (err error) {
	err = a.requireIndexFailed(index)
	if err == nil {
		a.data[index] = val
	}
	return
}

func (a *Array) IsEmpty() bool {
	return a.size == 0
}

//FIXME Contains
//查找数组中是否有此元素
func (a *Array) Contains(val interface{}) bool {
	for ele := range a.data {
		if ele == val {
			return true
		}
	}
	return false
}

//FIXME Find
//查找数组元素的索引，-1为没找到
func (a *Array) Find(val interface{}) int {
	for index, item := range a.data {
		if item == val {
			return index
		}
	}
	return -1
}

//FIXME Add
//在第index的位置插入一个新元素
func (a *Array) Add(index int, val interface{}) (err error) {
	err = a.requireIndexFailed(index)

	if a.size == len(a.data) {
		a.resize(2 * len(a.data))
	}

	for i := a.size - 1; i >= index; i-- {
		a.data[i+1] = a.data[i]
	}

	a.data[index] = val
	a.size++

	return
}

//FIXME AddFirst
//所有元素前添加新元素
func (a *Array) AddFirst(val interface{}) (err error) {
	err = a.Add(0, val)
	return
}

//FIXME AddLast
//所有元素后添加新元素
func (a *Array) AddLast(val interface{}) (err error) {
	err = a.Add(a.size, val)
	return
}

//FIXME Remove
//从数组删除index索引的元素，返回删除的元素值
func (a *Array) Remove(index int) (val interface{}, err error) {
	err = a.requireIndexFailed(index)

	val = a.data[index]
	for i := index + 1; i < a.size; i++ {
		a.data[i-1] = a.data[i]
	}
	a.size--
	a.data[a.size] = nil

	//防止resize频繁，在1/4时才减容
	//不等于0防止数组只有一个元素的情况
	if a.size == len(a.data)/4 && len(a.data)/2 != 0 {
		a.resize(len(a.data) / 2)
	}

	return
}

//删除数组第一个元素
func (a *Array) RemoveFirst() (interface{}, error) {
	return a.Remove(0)
}

//删除数组最后一个元素
func (a *Array) RemoveLast() (interface{}, error) {
	return a.Remove(a.size - 1)
}

//FIXME RemoveElement
//删除数组中的某个元素
func (a *Array) RemoveElement(ele interface{}) {
	if index := a.Find(ele); index != -1 {
		a.Remove(index)
	}
}

func (a *Array) String() string {
	buffer := bytes.Buffer{}
	desc := fmt.Sprintf("size = %d, cap = %d\n", a.size, a.cap)
	buffer.WriteString(desc)
	buffer.WriteByte('[')

	for index, item := range a.data {
		if index < a.size {
			buffer.WriteString(fmt.Sprintf("%v", item))
			if index != a.size-1 {
				buffer.WriteString(", ")
			}
		}
	}

	buffer.WriteByte(']')
	buffer.WriteByte('\n')

	return buffer.String()
}

//FIXME resize
//处理数组容量
func (a *Array) resize(cap int) {
	newData := make([]interface{}, cap)

	for i := 0; i < a.size; i++ {
		newData[i] = a.data[i]
	}
	a.cap = len(newData)
	a.data = newData
}

//FIXME requireIndexFailed
//判断索引index是否超出数组元素数或小于0个数组元素数
func (a *Array) requireIndexFailed(index int) (err error) {
	if index < 0 || index >= a.size {
		return errors.New("func params index < 0 or index <= size")
	}
	return nil
}
